x-pack/plugin/search-business-rules/src/main/java/org/apache/lucene/search/CappedScoreQuery.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.apache.lucene.search;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.util.Bits;
+
+/**
+ * A query that wraps another query and ensures scores do not exceed a maximum value
+ */
+public final class CappedScoreQuery extends Query {
```

Comment:
Sorry if I'm missing something obvious, I'm curious why we cap the scores as opposed to normalizing them to lie within a bounded range?

Reply:
Ideally we'd normalize rather than cap but in practice that would require us to change the way we collect results - we'd have to buffer matches to then release them to downstream collectors (hits, topHits, sampler agg...) with the rebased scores.
The CappedScoreQuery may have little use outside of the PinnedQuery use case - it's just a way for us to ensure organic scores never stray into the range of pinned scores. Once a pinned score is produced for a doc and collected by downstream priority queues we need to be sure a subsequent "organically-scored" doc does not exceed this score.

Reply:
I think it's also a nice property that non-pinned hits will usually get the same score whether the query is wrapped under a pinned query or not (the cap we set on scores is very high).

Reply:
Thanks for the context!